Mining Rework Part 1: First of his Name

BogCreature opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
In a vain attempt to make mining less boring and actually require more player engagement I have cobbled together the groundwork to a sizable reworking of the mining system from that thing you send out that one random crew member to go do for a couple minutes to a team-based activity focused around the use of specialized equipment.

The gameplay loop for mining I envision within my mind's eye is one where most members of the ship's crew, even down to the engineers and medics, have to cooperate in order to properly secure large amounts of ore from dangerous environments, with most other non-mining ships needing to at least partially rely on mining ships to supply them with minerals.

This PR specifically adds; ore veins (they use the same sprites as geysers pending a better one) that spawn on lava, sand, rock, and ice planets currently and at varying levels of difficulty and yeild, a drill with which to mine said veins, a scanner to find the aforementioned veins, and an outpost mission that is essentially just a mission to go drill ore veins but with money as the reward instead of ore. Oh and I also cut surface mineral spawn rates roughly in half pending a better gauge of the balance involved there.

Here's the harrowing story of one such man who tried to use this drill with no armaments and no crew:
![screenshot1](https://github.com/shiptest-ss13/Shiptest/assets/112462947/1a737ff6-5372-46ab-bda3-72363dbfe94d)
![screenshot2](https://github.com/shiptest-ss13/Shiptest/assets/112462947/fb9452b8-c6cd-4e9e-8986-59bbb272bdf3)
![screenshot3](https://github.com/shiptest-ss13/Shiptest/assets/112462947/812c4c8f-d3f7-4bde-a119-e8e6f35790e8)


<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Mining in its current state is just kinda bad. It's boring to actually focus on doing and way too easy to just drown your ship in minerals. This PR would at bare minimum partially fix that. That's not to say it's perfect and doesn't need to be balanced and polished to hell and back, but it's a good first step I feel.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog

:cl:
add: Ore veins and a drill to mine them with
add: Mining based missions
tweak: Lowered the spawn rate of some other mob spawners like tendrils
tweak: Added the ability for mob spawners to spawn mobs a distance from their source
balance: Cut most ore sources in half and tweaked the spawn weights
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

BogCreature committed:
Drill foundation

BogCreature committed:
I THOUGHT I KILLED YOU!!!

BogCreature committed:
Animations and ore vein file

BogCreature committed:
Spawner changes and polish

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Malfunctions part 1

BogCreature committed:
Hate

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Beware: Nearing Completion

BogCreature committed:
Mission nightmare begins

BogCreature committed:
Bugged

BogCreature committed:
I'm going to puke blood

Vile pinpointer code

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Judgement Day: TestMerge-ocalypse

BogCreature commented:
Of course I accidentally screenshot a spelling mistake.....

BogCreature committed:
Sand & ice planet spawn tweaks

BogCreature commented:
Ough, I forgot to mark this as a draft!

BogCreature committed:
I used the bitflags wrong

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
515 funny update

BogCreature committed:
I thought of a better place to put it

BogCreature committed:
Refines some stuff

Cleaned up some shitty code and improved the repair process

BogCreature committed:
More refining

BogCreature committed:
Spawner bugfixes

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Reworks the wave spawning to runtime less

BogCreature committed:
Last bit of stuff I can think of

BogCreature commented:
I can't really think of ways to refine any more of this stuff outside balance testing so I'm gonna cautiously say it's ready for review

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Fixes that absolutely vile pinpointer runtime

thgvr commented on code/datums/mapgen/planetary/LavaGenerator.dm:
```diff
@@ -112,9 +115,9 @@
-		/obj/structure/spawner/lavaland/low_threat = 12,
-		/obj/structure/spawner/lavaland/medium_threat = 4,
-		/obj/structure/spawner/lavaland/high_threat = 2,
+		/obj/structure/spawner/lavaland/low_threat = 8,
```
Any reason for the weight changes specifically?

BogCreature replied:
Yeah, I turned down tendril weight a bit because the amount of mobs spawning entities per planet just felt a bit excessive at the previous weights and with ore veins. Though these values were more just me probing the waters with what would be better during local testing.

thgvr replied:
Seems fine at least - easy to tweak, anyways.

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,337 @@
+	if(panel_open && component_parts)
+		. += display_parts(user, TRUE)
+	if(cell.charge < power_cost*5)
+		. += "<spawn class='notice'>The lower power light is blinking."
```
```suggestion
		. += "<spawn class='notice'>The low power light is blinking."
```

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,337 @@
+//Instead of being qdeled the drill requires mildly expensive repairs to use again
+/obj/machinery/drill/deconstruct(disassembled)
+	if(active && mining)
+		say("Drill integrity failure, comencing emergency shutdown procedure.")
```
```suggestion
		say("Drill integrity failure. Engaging emergency shutdown procedure.")
```

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,337 @@
+
+//The RPED sort of trivializes a good deal of the malfunction mechancis, as such it will not be allowed to work
+/obj/machinery/drill/exchange_parts(mob/user, obj/item/storage/part_replacer/W)
+	to_chat(user, "<span calss='notice'>[W] does not seem to work on [src], it might require more delecitate parts replacement.")
```
```suggestion
	to_chat(user, "<span class='notice'>[W] does not seem to work on [src], it might require more delicate part manipulation.")
```

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,337 @@
+		say("Please resolve existing malfunction before continuing mining operations.")
+		return
+	if(!mining)
+		to_chat(user, "<span class='notice'>[src] isn't sercured over an ore vein!</span>")
```
```suggestion
		to_chat(user, "<span class='notice'>[src] isn't secured over an ore vein!</span>")
```

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,337 @@
+		start_mining()
+		return
+	else
+		to_chat(user, "<span class='notice'>[src] is currently busy, wait till it's done!</span>")
```
```suggestion
		to_chat(user, "<span class='notice'>[src] is currently busy, wait until it's done!</span>")
```

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,337 @@
+	for(var/obj/item/stock_parts/capacitor/capacitor in component_parts)
+		power_use = power_cost/capacitor.rating
+	if(cell.charge < power_use)
+		say("Error: Internal cell charge deplted")
```
```suggestion
		say("Error: Internal cell charge depleted")
```

thgvr commented on code/modules/overmap/missions/drill_mission.dm:
```diff
@@ -0,0 +1,99 @@
+/datum/mission/drill
+	name = "Class 1 core sample mission"
+	desc = "We require geological information from one of the neighboring planetoids . \
+			Please anchor the drill in place and defend it until it has gathered enough smaples.\
```
```suggestion
			Please anchor the drill in place and defend it until it has gathered enough samples.\
```

thgvr commented on code/modules/overmap/missions/drill_mission.dm:
```diff
@@ -0,0 +1,99 @@
+	name = "Class 1 core sample mission"
+	desc = "We require geological information from one of the neighboring planetoids . \
+			Please anchor the drill in place and defend it until it has gathered enough smaples.\
+			Operation of the core sampling drill is extremely dangerous, use caution. "
```
```suggestion
			Operation of the core sampling drill is extremely dangerous, caution is advised. "
```

thgvr reviewed (changes requested):
Code looks fine on cursory review, but will need a once-or-twice look over from tmt before merging. I picked out some errors I found at least.

BogCreature committed:
I'm genuinely surprised I'm even literate

Co-authored-by: thgvr <81882910+thgvr@users.noreply.github.com>
Signed-off-by: BogCreature <112462947+BogCreature@users.noreply.github.com>

BogCreature committed:
Smaples...

Co-authored-by: thgvr <81882910+thgvr@users.noreply.github.com>
Signed-off-by: BogCreature <112462947+BogCreature@users.noreply.github.com>

BogCreature committed:
Spawning scales with player count

Spawns roughly 1 enemy per 2 players within range of any spawner with a max range higher than 1 tile

BogCreature committed:
Fixes a minor yet annoying bug

BogCreature committed:
Merge branch 'TestTest' of https://github.com/BogCreature/Shiptest into TestTest

BogCreature committed:
Forgot to remove this before pushing, my bad

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Infinite mining loot fix

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,338 @@
+	if(cell.charge < power_cost*5)
+		. += "<spawn class='notice'>The low power light is blinking."
+	switch(malfunction)
+		if(4)
```
I see this starts at four. Can you add some examine text here for stock part malfunctions?

thgvr commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,338 @@
+		return TRUE
+	if(panel_open) //All malfunction repair and maintenance actions are handled under here
+		var/list/needed_parts = list(/obj/item/stock_parts/scanning_module,/obj/item/stock_parts/micro_laser,/obj/item/stock_parts/capacitor)
+		if(is_type_in_list(tool,needed_parts))
```
This needs some to_chat feedback for stock part insertion.

thgvr reviewed (changes requested):
Few more things I noticed

BogCreature committed:
Vein time and malfunction tweak

BogCreature committed:
FUCK!!!

BogCreature committed:
I hate this

BogCreature committed:
Forgot to change these

BogCreature committed:
Fix to stock part hell

BogCreature commented:
Right there should be a bit more player feedback and the stock part replacement code should be notably less awful

BogCreature committed:
Temporarily buffed drill output overall

Holdover until I have a better method of ore distribution

BogCreature committed:
Forgot I moved this around on the main drill

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -22,14 +29,29 @@
+		Destroy(TRUE,FALSE)
+		return
+	if(spawning_paused)
+		sleep(wave_downtime)
```
this technically works but it's very not good in terms of lag
ideally you'd stop it from processing and set a timer that re-adds it to SSprocessing after that timer has elapsed

BogCreature replied:
I'll be bluntly honest I thought I'd already deleted this segment of code, but I did take your suggestion and handle it via SSprocessing

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
 
 /datum/component/spawner/proc/try_spawn_mob()
 	var/atom/P = parent
+	var/turf/spot = P.loc
```
```suggestion
	var/turf/spot = get_turf(P)
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
+	spawn_multiplier = clamp(spawn_multiplier, 1, max_mobs - spawned_mobs.len)
+	for(spawn_multiplier, spawn_multiplier > 0, spawn_multiplier--)
+		if(spawn_distance_max > 1)
+			spot = pick(turf_peel(spawn_distance_max, spawn_distance_min, P.loc, view_based = TRUE))
```
```suggestion
			spot = pick(turf_peel(spawn_distance_max, spawn_distance_min, spot, view_based = TRUE))
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
+		if(spawn_distance_max > 1)
+			spot = pick(turf_peel(spawn_distance_max, spawn_distance_min, P.loc, view_based = TRUE))
+			if(!spot)
+				spot = pick(circlerangeturfs(P.loc, spawn_distance_max))
```
```suggestion
				spot = pick(circlerangeturfs(spot, spawn_distance_max))
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
+	if(wave_timer && world.time > wave_timer)
+		spawning_paused = TRUE
+		wave_timer = null
+		return 0
```
```suggestion
	if(wave_length)
		if(!wave_timer)
			COOLDOWN_START(src, wave_timer, wave_length)
		if(wave_timer && COOLDOWN_FINISHED(src, wave_timer))
			spawning_paused = TRUE
			COOLDOWN_RESET(src, wave_timer)
			return
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
+			downtime_timer = null
+			return 0
+		else
+			return 0
```
```suggestion
	if(spawning_paused)
		if(!downtime_timer)
			COOLDOWN_START(src, downtime_timer, wave_downtime)
		if(COOLDOWN_FINISHED(src, downtime_timer))
			spawning_paused = FALSE
			COOLDOWN_RESET(src, downtime_timer)
		return
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
 	if(spawned_mobs.len >= max_mobs)
 		return 0
 	if(spawn_delay > world.time)
 		return 0
```
```suggestion
	if(length(spawned_mobs) >= max_mobs)
		return
	if(COOLDOWN_FINISHED(src, spawn_delay))
		return
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
 		return 0
+	//Avoid using this with spawners that add this component on initialize
+	//It causes numerous runtime errors during planet generation
 	spawn_delay = world.time + spawn_time
```
```suggestion
	COOLDOWN_START(src, spawn_delay, spawn_time)
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -41,17 +63,54 @@
+		for(var/obj/mecha/mechs in range(spawn_distance_max, P.loc))
+			player_count++
+		if(player_count > 3)
+			spawn_multiplier = round(player_count/2)
```
```suggestion
		for(var/mob/player as anything in GLOB.player_list)
			if(!isliving(player))
				continue
			if(player.stat != CONSCIOUS)
				continue
			if(get_dist(get_turf(player), spot) > spawn_distance_max)
				continue
			player_count++
		if(player_count > 3)
			spawn_multiplier = round(player_count/2)
```

BogCreature committed:
Cooldown stuffs

Co-authored-by: Mark Suckerberg <mark@suckerberg.gay>
Signed-off-by: BogCreature <112462947+BogCreature@users.noreply.github.com>

DireWaff commented:
Thumpers from FireFall... Thumbs up, chief.

BogCreature committed:
Spawner waves improved

Switches spawner waves to temporarily stopping processing instead of the previous method

BogCreature committed:
Drill anti cheese measure

Also buffed ore output a bit

thgvr commented on code/modules/overmap/missions/drill_mission.dm:
```diff
@@ -0,0 +1,108 @@
+	weight = 8
+
+	var/obj/machinery/drill/mission/sampler
+	var/num_wanted = 8
```
Can you try lowering this to something like 6? Maybe 5?

BogCreature replied:
It has been done

BogCreature replied:
(I meant to do this back when I adjusted the amount of charges per vein but I forgot to)

BogCreature committed:
Balance tweaks

BogCreature committed:
I like these numbers a bit better

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Drill mission progress fix

BogCreature committed:
Finer control of veins

BogCreature committed:
No more dwarf legion farming

BogCreature committed:
runtime fix

BogCreature committed:
Rebalances and a bugfix

BogCreature committed:
Merge remote-tracking branch 'upstream/master' into TestTest

BogCreature committed:
Off button update

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -22,36 +29,92 @@
 
 /datum/component/spawner/process()
+	if(!parent) //Sanity check for instances where the spawner may be sleeping while the parent is destroyed
+		Destroy(TRUE,FALSE)
```
```suggestion
		qdel(src)
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -22,36 +29,92 @@
+			player_count++
+		if(player_count > 3)
+			spawn_multiplier = round(player_count/2)
+	spawn_multiplier = clamp(spawn_multiplier, 1, max_mobs - spawned_mobs.len)
```
```suggestion
	spawn_multiplier = clamp(spawn_multiplier, 1, max_mobs - length(spawned_mobs))
```

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -22,36 +29,92 @@
+		if(player_count > 3)
+			spawn_multiplier = round(player_count/2)
+	spawn_multiplier = clamp(spawn_multiplier, 1, max_mobs - spawned_mobs.len)
+	for(spawn_multiplier, spawn_multiplier > 0, spawn_multiplier--)
```
```suggestion
	for(var/mob_index in 1 to spawn_multiplier)
```
same difference, just a style thing

MarkSuckerberg commented on code/datums/components/spawner.dm:
```diff
@@ -22,36 +29,92 @@
+		if(spawn_distance_max > 1)
+			var/origin = spot
+			var/list/peel = turf_peel(spawn_distance_max, spawn_distance_min, origin, view_based = TRUE)
+			if(peel.len>0)
```
```suggestion
			if(length(peel))
```

MarkSuckerberg commented on code/game/objects/items/pinpointer.dm:
```diff
@@ -63,7 +63,7 @@
 	if(!active)
 		return
-	if(!target)
+	if(!target || target.loc == null)
```
```suggestion
	if(!target?.loc)
```

MarkSuckerberg commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,379 @@
+			anchored = TRUE
+			update_icon_state()
+			return
+		else if(tool.use_tool(src, user, 30, volume=50))
```
```suggestion
		if(tool.use_tool(src, user, 30, volume=50))
```
since you return already in the if

MarkSuckerberg commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,207 @@
+	return..()
+
+/obj/structure/vein/proc/begin_spawning()
+	spawner = AddComponent(spawner_type, mob_types, spawn_time, faction, spawn_text, max_mobs, spawn_sound, spawn_distance_min, spawn_distance_max, wave_length, wave_downtime)
```
Usually we don't store components as variables-- anything we need to do interact with it, we do with signals. Might be useful to make a start_ and stop_spawning signal, or something, if you need to.

Just removing the component works too, though

tmtmtl30 replied:
yeah, storing a reference to a component is something you really shouldn't do, generally. it can be removed by a variety of things. you should probably set this reference to null in /obj/structure/vein/Destroy(); depending on the exact order things get resolved in, i can see it causing harddels.

BogCreature replied:
Apologies for not getting to this sooner, I got distracted by a number of things and forget about it. I'll try and address this sometime in the next couple days.

MarkSuckerberg commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,207 @@
+/obj/structure/vein/proc/begin_spawning()
+	spawner = AddComponent(spawner_type, mob_types, spawn_time, faction, spawn_text, max_mobs, spawn_sound, spawn_distance_min, spawn_distance_max, wave_length, wave_downtime)
+
+/obj/structure/vein/proc/toggle_spawning()
```
Incomplete or accidentally left in?

BogCreature replied:
Accidentally left in

MarkSuckerberg commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,207 @@
+				ore_type_amount = rand(4,6)
+			else
+				ore_type_amount = 1
+		for(ore_type_amount, ore_type_amount>0, ore_type_amount--)
```
```suggestion
		for(var/ore_count in 1 to ore_type_amount)
```

MarkSuckerberg commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,207 @@
+				ore_type_amount = 1
+		for(ore_type_amount, ore_type_amount>0, ore_type_amount--)
+			var/picked
+			picked = pickweight(ore_list)
```
```suggestion
			var/picked = pickweight(ore_list)
```

MarkSuckerberg commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,207 @@
+
+/obj/structure/vein/Destroy()
+	. = ..()
+	GLOB.ore_veins -= src
```
```suggestion
	GLOB.ore_veins -= src
	return ..()
```

MarkSuckerberg commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,207 @@
+	class = vein_class
+	for(class, class>0, class--)
+		var/picked
+		picked = pick(vein_contents)
```
```suggestion
	for(var/vein_content_count in 1 to class)
		var/picked = pick(vein_contents)
```

MarkSuckerberg reviewed (changes requested):
Sorry this is taking so long :c

BogCreature committed:
Apply suggestions from code review

Co-authored-by: Mark Suckerberg <mark@suckerberg.gay>
Signed-off-by: BogCreature <112462947+BogCreature@users.noreply.github.com>

BogCreature committed:
Apply suggestions from code review

Co-authored-by: Mark Suckerberg <mark@suckerberg.gay>
Signed-off-by: BogCreature <112462947+BogCreature@users.noreply.github.com>

BogCreature committed:
Ough

BogCreature committed:
Apply suggestions from code review

Co-authored-by: Mark Suckerberg <mark@suckerberg.gay>
Signed-off-by: BogCreature <112462947+BogCreature@users.noreply.github.com>

thgvr commented:
I'm tempted to merge this but I'll bother tmt to go over it, bit out of my range

tmtmtl30 commented on code/modules/mining/equipment/mineral_scanner.dm:
```diff
@@ -37,7 +40,7 @@
 
 /obj/item/t_scanner/adv_mining_scanner
-	desc = "A scanner that automatically checks surrounding rock for useful minerals; it can also be used to stop gibtonite detonations. This one has an extended range.\nIt has a speaker that can be toggled with <b>alt+click</b>"
+	desc = "A scanner that automatically checks surrounding rock for useful minerals; it can also be used to stop gibtonite detonations. This one has an extended range and a ground penetrating scanner.\nIt has a speaker that can be toggled with <b>alt+click</b>"
```
not sure why the desc is changed, since its behavior stays the same?

tmtmtl30 commented on code/modules/overmap/missions/drill_mission.dm:
```diff
@@ -0,0 +1,111 @@
+		return
+	. = ..()
+
+/obj/machinery/drill/mission/mine()
```
feels like there's too much duplication of the parent proc here. maybe have a proc "mine_success()" that on /obj/machinery/drill spawns ore and is overridden in /obj/machinery/drill/mission to increment num_current?

tmtmtl30 commented on code/modules/overmap/missions/drill_mission.dm:
```diff
@@ -0,0 +1,111 @@
+
+/obj/machinery/drill/mission/start_mining()
+	if(mining.vein_class < mission_class && mining)
+		say("[src] requires at least a class [mission_class] vein or higher.")
```
feels silly for the drill to talk about itself in third person. just make this a message in the "Error:" format used by the parent class' error/malfunction messages

tmtmtl30 commented:
currently reviewing this; will finish looking at it tomorrow

tmtmtl30 commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,379 @@
+		if(metal_attached == METAL_SECURED && tool.tool_behaviour == TOOL_WELDER)
+			if(tool.use_tool(src, user, 30, volume=50))
+				to_chat(user, "<span class='notice'>You weld the new plating onto the [src], successfully repairing it.")
+				metal_attached = null
```
i'd make this set to a define equal to zero (which should also be the default state), to make it clear that falsey metal_attached is intended behavior

tmtmtl30 commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,379 @@
+		. += display_parts(user, TRUE)
+	if(cell.charge < power_cost*5)
+		. += "<spawn class='notice'>The low power light is blinking.</span>"
+	switch(malfunction)
```
these should use the defines

tmtmtl30 commented on code/modules/mining/equipment/mineral_scanner.dm:
```diff
@@ -100,3 +103,128 @@
+	var/turf/here = get_turf(src)
+	var/located_dist
+	var/obj/structure/located_vein
+	for(var/obj/structure/I in GLOB.ore_veins)
```
you can write "for(var/obj/structure/I as anything in GLOB.ore_veins)" if you know that the list will always be populated with instances of /obj/structure, as by default typed loops like this will run a typecheck on every member of the target list before running the loop body on them.

(also, it's generally good practice to use the least general type that you're looking for when you don't intend to reuse this code, for clarity -- /obj/structure/vein, instead of just /obj/structure)

using a global list for this sort of thing is good, though -- if it wasn't here, i'd be worried about players lagging the server by spam-clicking on the item while mashing Z, as "get all X objects within Y range" can be fairly expensive; it's something byond is bad at.

BogCreature replied:
>using a global list for this sort of thing is good, though -- if it wasn't here, i'd be worried about players lagging the server by spam-clicking on the item while mashing Z, as "get all X objects within Y range" can be fairly expensive; it's something byond is bad at.

That's really good to know, optimization's something I don't have a terribly good grasp of and I feel like that's a mistake I would've ended up making in the future

tmtmtl30 commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,202 @@
+	class = vein_class
+	for(var/vein_content_count in 1 to class)
+		var/picked = pick(vein_contents)
+		new picked(pick(get_adjacent_open_turfs(current)),round(rand(drop_rate_amount_min,drop_rate_amount_max)*multiplier))
```
note that get_adjacent_open_turfs might return an empty list -- probably not a great idea to be spawning the ore in nullspace? if you don't find a turf, i'd just fallback to the current one, or something

tmtmtl30 commented on code/modules/mining/ore_veins.dm:
```diff
@@ -0,0 +1,202 @@
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+
+	var/mining_charges = 6
+	//Classification of the quality of possible ores within a vein, used to determine difficulty
```
since it looks like the vein_class is expected to lie within 1 and 3 (inclusive), i'd add a comment making that clear

tmtmtl30 commented on code/datums/components/spawner.dm:
```diff
@@ -7,9 +7,16 @@
+	var/wave_length //Average time until break in spawning
+	var/wave_downtime //Average time until spawning starts again
+	var/wave_timer
+	var/downtime_timer
```
unused, and should be removed

MarkSuckerberg commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,379 @@
+			to_chat(user, "<span class='notice'>You unsecure the [src] from the ore vein.</span>")
+			playsound(src, 'sound/items/deconstruct.ogg', 50, TRUE)
+			anchored = FALSE
+			if(mining.spawner)
```
```suggestion
			if(mining?.spawner)
```
This is a solution to a runtime on live, but it may be a bigger issue of mining being null when it shouldn't be. The runtime in question:

```
Cannot read null.spawner : 5
 | attackby (/obj/machinery/drill/attackby) : 5
 | - file: code/modules/mining/drill.dm,148
 | - usr: oc: the iron sand (147,238,5) (/turf/open/floor/plating/asteroid/rockplanet/lit)
 | - src: the core sampling research dri... (/obj/machinery/drill/mission)
 | - loc: the snow (148,238,5) (/turf/open/floor/plating/asteroid/snow/lit/rockplanet)
 | - stack: 
 |   - the core sampling research dri... (/obj/machinery/drill/mission): attackby(the wrench (/obj/item/wrench), Beedost Oktba (/mob/living/carbon/human), "icon-x=17;icon-y=14;left=1;but...")
 |   - the wrench (/obj/item/wrench): melee attack chain(Beedost Oktba (/mob/living/carbon/human), the core sampling research dri... (/obj/machinery/drill/mission), "icon-x=17;icon-y=14;left=1;but...")
 |   - Beedost Oktba (/mob/living/carbon/human): ClickOn(the core sampling research dri... (/obj/machinery/drill/mission), "icon-x=17;icon-y=14;left=1;but...")
 |   - the core sampling research dri... (/obj/machinery/drill/mission): Click(the snow (148,238,5) (/turf/open/floor/plating/asteroid/snow/lit/rockplanet), "mapwindow.map", "icon-x=17;icon-y=14;left=1;but...")

BogCreature replied:
Ough, I'm not sure how I never found this one while I was testing it initially, it's almost assuredly just null because the vein's been deleted so I don't think it's a more widespread issue, but it's such an insanely common occurrence I'm not sure how I never stumbled into it at least once.

BogCreature committed:
Null mining variable fix

Co-authored-by: Mark Suckerberg <mark@suckerberg.gay>
Signed-off-by: BogCreature <112462947+BogCreature@users.noreply.github.com>

tmtmtl30 commented on code/datums/components/spawner.dm:
```diff
@@ -22,36 +29,92 @@
+				continue
+			if(player.stat != CONSCIOUS)
+				continue
+			if(get_dist(get_turf(player), spot) > spawn_distance_max)
```
get_dist calculates 3D distance (that is, including across z-levels). i'd note that GLOB.player_list includes observers etc. but the isliving() check takes care of those

tmtmtl30 commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -582,7 +590,7 @@
 
 /mob/living/simple_animal/hostile/proc/ListTargetsLazy(virtual_z)//Step 1, find out what we can see
-	var/static/hostile_machines = typecacheof(list(/obj/machinery/porta_turret, /obj/mecha)) //WS - add spacepod
+	var/static/hostile_machines = typecacheof(list(/obj/machinery/porta_turret, /obj/mecha, /obj/machinery/drill)) //WS - add spacepod
```
changing this doesn't really do anything because of the way this list is used, but for consistency it's probably fine

tmtmtl30 reviewed (changes requested):
lots of nitpicks and concerns

BogCreature committed:
First batch of review fixes

Clears out an unused variable, adjusts some drill defines, and fixes a description I accidentally changed

BogCreature committed:
Second batch of review fixes

Enhanced ore vein comments, slightly improved scanner for loop, and bizzare and gormless mission drill writing improved

west3436 commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,380 @@
+			update_overlays()
+		else
+			to_chat(user, "<span class='notice'>You cancel the manual shutoff process.</span>")
+			return
```
Remove trailing return

west3436 commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,380 @@
+		return
+	else
+		to_chat(user, "<span class='notice'>[src] is currently busy, wait until it's done!</span>")
+		return
```
Trailing return

west3436 commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,380 @@
+		say("Estimated time until vein depletion: [time2text(eta,"mm:ss")].")
+		update_icon_state()
+		update_overlays()
+		return
```
Trailing return

west3436 commented on code/modules/mining/drill.dm:
```diff
@@ -0,0 +1,380 @@
+			return
+		if(MALF_CALIBRATE)
+			say("Malfunction: Drill laser calibrations out of alignment, please recalibrate before continuing.")
+			return
```
Trailing return

BogCreature committed:
Divides mine() into 2 procs and fixes a rare bug

mine() is now mine() and mine_success() so there's less oddity with mission drills and ore veins now have a fallback plan for if there's no open adjacent turfs

BogCreature committed:
Obliterates trailing returns

BogCreature committed:
Signalizes the ore vein spawning component

Does away with the spawner component reference and converts it to use signals instead

BogCreature committed:
Stops an extremely rare bug

Stops an extremely rare bug where if someone is on roughly the same position of an adjacent zlevel they would nudge the mob spawn modifier on the spawners
